<a href="https://colab.research.google.com/github/gimbabheaven/2023-Basic_Python/blob/master/07_%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EB%B0%98%EB%B3%B5%EB%AC%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 반복문

- 1회 이상 반복해서 수행하는 일을 반복문으로 통해 쉽게 해결할 수 있습니다.
- 반드시 list, tuple, dict, set 등 집합에 대한 순회를 돌며 일을 처리할 때 많이 쓰입니다.
- list, tuple, dict, set, 문자열까지 모두 **순회가능한(iterable) 객체** 입니다.
- 순회가능한(iterable) 객체는 반복문을 통해 순회할 수 있습니다.

In [ ]:
mylist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

mylist에 들어 있는 모든 값들을 출력하려고 한다면?

In [ ]:
print(mylist[0])
print(mylist[1])
print(mylist[2])
print('...')
print(mylist[8])
print(mylist[9])

1
2
3
...
9
10


## for 와 in 구문 (반복문)

**[기본 문법]**<br>
for 하나씩 꺼내올 때 변수 in [꺼내올 집합]:<br/>
(indent)
* list, tuple, set, dictionary, 문자열 형태 모두 가능
* range와 결합하여 사용 가능

### 리스트(list)

In [ ]:
mylist = [1, 2, 3, 4, 5]

In [ ]:
# 코드를 입력해 주세요
for i in mylist:
  print(i)

1
2
3
4
5


<p><strong>[출력 결과]</strong></p><pre>1
2
3
4
5
</pre>

### 튜플(tuple)

In [ ]:
for i in (1, 2, 3, 4, 5):
    print(i)

1
2
3
4
5


### tuple + list

tuple을 전체로 받아주는 경우

In [ ]:
person = ('제이콥스', 10)
print(person)
print(person[0])
print(person[1])

('제이콥스', 10)
제이콥스
10


tuple의 요소를 개별로 받아주는 경우

In [ ]:
name, age = ('제이콥스', 10)
print(name)
print(age)

제이콥스
10


반복문에서의 응용

In [ ]:
mytuplelist = [('제이콥스', 10), ('피터', 20), ('타이거', 30)]

for mytuple in mytuplelist:
    print(mytuple[0], mytuple[1])

제이콥스 10
피터 20
타이거 30


In [ ]:
mytuplelist = [('제이콥스', 10), ('피터', 20), ('타이거', 30)]

for name, age in mytuplelist:
    print(name, age)

제이콥스 10
피터 20
타이거 30


### 딕셔너리(dictionary)

In [ ]:
mydict = {'헐크': 50, '아이언맨': 60, '펭수': 70}

In [ ]:
mydict.keys()

dict_keys(['헐크', '아이언맨', '펭수'])

In [16]:
for key in mydict.keys():
    print(key)

헐크
아이언맨
펭수


In [ ]:
for value in mydict.values():
    print(value)

50
60
70


In [ ]:
mydict.items()

dict_items([('헐크', 50), ('아이언맨', 60), ('펭수', 70)])

In [17]:
for name, age in mydict.items():
    print(name, age)

헐크 50
아이언맨 60
펭수 70


### 문자열(str)

In [18]:
for c in "Hello":
    print(c)

H
e
l
l
o


### range()

- `range()` 함수는 별도의 list, tuple 생성 없이 range() 에서 정의한 범위를 반복는데 활용할 수 있습니다.
- `range(start, stop, step)` 형식을 사용합니다.

`stop`: 단일 값을 지정하는 경우

In [19]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


`start`, `stop`: 두 개의 값을 지정한 경우

In [20]:
for i in range(2, 9):
    print(i)

2
3
4
5
6
7
8


`start`, `stop`, `step`: 세 개의 값을 지정한 경우

In [21]:
for i in range(1, 10, 2):
    print(i)

1
3
5
7
9


## 반복문의 중첩

In [22]:
for i in range(1, 4):
    for j in range(1, 4):
        print(f'(i={i}) + (j={j}) = {i * j}')
    print('===')

(i=1) + (j=1) = 1
(i=1) + (j=2) = 2
(i=1) + (j=3) = 3
===
(i=2) + (j=1) = 2
(i=2) + (j=2) = 4
(i=2) + (j=3) = 6
===
(i=3) + (j=1) = 3
(i=3) + (j=2) = 6
(i=3) + (j=3) = 9
===


## 제어문

### continue

- 반복문 내부에서 `continue` 구문은 해당 루프(loop)를 건너뛰게 합니다.
- `continue` 라는 구문을 만나면, 반복문에서 `continue` 아래 작성된 코드는 **실행되지 않고 건너뜁니다.**

In [23]:
mylist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

**반복문**과 **조건문** 그리고 `continue`를 활용하여 **짝수만 출력**해 주세요

In [24]:
# 코드를 입력해 주세요
for i in mylist:
  if i % 2 == 1:
    continue
    print(i)

<p><strong>[출력 결과]</strong></p><pre>2
4
6
8
10
</pre>

### break

- **break** 구문을 만나면, **반복 루프(loop)는 즉시 종료**됩니다.

In [25]:
mylist = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

`break`를 사용하여 i가 6 이상이면 STOP

In [26]:
for i in mylist:
    if i >= 6:
        break
    print(i)

1
2
3
4
5


<p><strong>[출력 결과]</strong></p><pre>1
2
3
4
5
</pre>

### break, continue 차이

In [27]:
for i in range(10):
    if i == 5:
        break
    print(i)

0
1
2
3
4


In [28]:
for i in range(10):
    if i == 5:
        continue
    print(i)

0
1
2
3
4
6
7
8
9


### 연습문제

- 구구단을 반복문을 사용하여 만들어 보세요!
- 구구단은 2단 부터 9단까지 있습니다.

In [30]:
# 코드를 입력해 주세요
for i in range(2, 10):
  for j in range(1, 10):
    print(i, "X", j, " = ", i*j)
  print()

2 X 1  =  2
2 X 2  =  4
2 X 3  =  6
2 X 4  =  8
2 X 5  =  10
2 X 6  =  12
2 X 7  =  14
2 X 8  =  16
2 X 9  =  18

3 X 1  =  3
3 X 2  =  6
3 X 3  =  9
3 X 4  =  12
3 X 5  =  15
3 X 6  =  18
3 X 7  =  21
3 X 8  =  24
3 X 9  =  27

4 X 1  =  4
4 X 2  =  8
4 X 3  =  12
4 X 4  =  16
4 X 5  =  20
4 X 6  =  24
4 X 7  =  28
4 X 8  =  32
4 X 9  =  36

5 X 1  =  5
5 X 2  =  10
5 X 3  =  15
5 X 4  =  20
5 X 5  =  25
5 X 6  =  30
5 X 7  =  35
5 X 8  =  40
5 X 9  =  45

6 X 1  =  6
6 X 2  =  12
6 X 3  =  18
6 X 4  =  24
6 X 5  =  30
6 X 6  =  36
6 X 7  =  42
6 X 8  =  48
6 X 9  =  54

7 X 1  =  7
7 X 2  =  14
7 X 3  =  21
7 X 4  =  28
7 X 5  =  35
7 X 6  =  42
7 X 7  =  49
7 X 8  =  56
7 X 9  =  63

8 X 1  =  8
8 X 2  =  16
8 X 3  =  24
8 X 4  =  32
8 X 5  =  40
8 X 6  =  48
8 X 7  =  56
8 X 8  =  64
8 X 9  =  72

9 X 1  =  9
9 X 2  =  18
9 X 3  =  27
9 X 4  =  36
9 X 5  =  45
9 X 6  =  54
9 X 7  =  63
9 X 8  =  72
9 X 9  =  81



<p><strong>[출력 결과]</strong></p><pre>2 X 1 = 2
2 X 2 = 4
2 X 3 = 6
2 X 4 = 8
2 X 5 = 10
2 X 6 = 12
2 X 7 = 14
2 X 8 = 16
2 X 9 = 18

3 X 1 = 3
3 X 2 = 6
3 X 3 = 9
3 X 4 = 12
3 X 5 = 15
3 X 6 = 18
3 X 7 = 21
3 X 8 = 24
3 X 9 = 27

4 X 1 = 4
4 X 2 = 8
4 X 3 = 12
4 X 4 = 16
4 X 5 = 20
4 X 6 = 24
4 X 7 = 28
4 X 8 = 32
4 X 9 = 36

5 X 1 = 5
5 X 2 = 10
5 X 3 = 15
5 X 4 = 20
5 X 5 = 25
5 X 6 = 30
5 X 7 = 35
5 X 8 = 40
5 X 9 = 45

6 X 1 = 6
6 X 2 = 12
6 X 3 = 18
6 X 4 = 24
6 X 5 = 30
6 X 6 = 36
6 X 7 = 42
6 X 8 = 48
6 X 9 = 54

7 X 1 = 7
7 X 2 = 14
7 X 3 = 21
7 X 4 = 28
7 X 5 = 35
7 X 6 = 42
7 X 7 = 49
7 X 8 = 56
7 X 9 = 63

8 X 1 = 8
8 X 2 = 16
8 X 3 = 24
8 X 4 = 32
8 X 5 = 40
8 X 6 = 48
8 X 7 = 56
8 X 8 = 64
8 X 9 = 72

9 X 1 = 9
9 X 2 = 18
9 X 3 = 27
9 X 4 = 36
9 X 5 = 45
9 X 6 = 54
9 X 7 = 63
9 X 8 = 72
9 X 9 = 81

</pre>

## While

- while 문은 while문과 함께 정의한 **조건이 참이 동안 반복 루프를 수행**합니다.

In [31]:
count = 5

while count > 0:
    print(count)
    count -= 1

5
4
3
2
1


많은 사용하는 방법 중 하나는 `while True`로 지정하여 무한 루프를 생성 후, **탈출 구문 루프 내에서 설정**하는 것입니다.

In [ ]:
count = 1

while True:
    print(count)
    count += 1
    # 탈출 구문(break)
    if count > 5:
        break

### 연습문제 1

사용자의 입력이 'MALE', 'FEMALE'인 경우는 while 문을 종료하고, 그렇지 않으면 "잘못된 입력 값입니다."를 출력하세요

- input() 함수로 사용자로부터 입력을 받습니다.

In [38]:
# 코드를 입력해 주세요
gender = ''

while(gender != 'MALE' and gender != 'FEMALE'):
    gender = input()
    if gender != 'MALE' and gender != 'FEMALE':
        print('잘못된 입력 값입니다.')

print('====\n프로그램 종료')

animal
잘못된 입력 값입니다.
male
잘못된 입력 값입니다.
MALE
====
프로그램 종료


<p><strong>[출력 결과]</strong></p><pre>animal
잘못된 입력 값입니다.
male
잘못된 입력 값입니다.
FEMALE
</pre>

### 연습 문제 2

1부터 50까지의 숫자 중에서 짝수의 합을 구하세요 (while 사용)

In [47]:
# 코드를 입력해 주세요
count = 1
total = 0

while (count <= 50):
  if (count % 2 == 0):
    total += count
    print('count : ', count, ", total : ", total)
  count += 1

count :  2 , total :  2
count :  4 , total :  6
count :  6 , total :  12
count :  8 , total :  20
count :  10 , total :  30
count :  12 , total :  42
count :  14 , total :  56
count :  16 , total :  72
count :  18 , total :  90
count :  20 , total :  110
count :  22 , total :  132
count :  24 , total :  156
count :  26 , total :  182
count :  28 , total :  210
count :  30 , total :  240
count :  32 , total :  272
count :  34 , total :  306
count :  36 , total :  342
count :  38 , total :  380
count :  40 , total :  420
count :  42 , total :  462
count :  44 , total :  506
count :  46 , total :  552
count :  48 , total :  600
count :  50 , total :  650


<p><strong>[출력 결과]</strong></p><pre>count: 2, total: 2
count: 4, total: 6
count: 6, total: 12
count: 8, total: 20
count: 10, total: 30
count: 12, total: 42
count: 14, total: 56
count: 16, total: 72
count: 18, total: 90
count: 20, total: 110
count: 22, total: 132
count: 24, total: 156
count: 26, total: 182
count: 28, total: 210
count: 30, total: 240
count: 32, total: 272
count: 34, total: 306
count: 36, total: 342
count: 38, total: 380
count: 40, total: 420
count: 42, total: 462
count: 44, total: 506
count: 46, total: 552
count: 48, total: 600
count: 50, total: 650
</pre>

### 연습 문제 3

**while**을 이용하여 다음과 같이 @ 찍기에 도전해 보세요!

In [48]:
# 코드를 입력해 주세요
num = 1
while num <= 5:
  print('@'*num)
  num += 1

@
@@
@@@
@@@@
@@@@@


<p><strong>[출력 결과]</strong></p><pre>@
@@
@@@
@@@@
@@@@@
</pre>

In [52]:
# 코드를 입력해 주세요
num = 1
while num<=5:
    if num <=3:
        print('@' * num)
    else:
        print('@' * (6-num))
    num += 1

@
@@
@@@
@@
@


<p><strong>[출력 결과]</strong></p><pre>@
@@
@@@
@@
@
</pre>

In [ ]:
# 코드를 입력해 주세요


<p><strong>[출력 결과]</strong></p><pre>  @
 @@@
@@@@@
 @@@
  @
</pre>